backends/p4tools/testgen/core/small_step/extern_stepper.cpp

```diff
@@ -0,0 +1,949 @@
+         */
+        {"packet_in.advance",
+         {"sizeInBits"},
+         [this](const IR::MethodCallExpression* call, const IR::Expression* /*receiver*/,
+                IR::ID& /*methodName*/, const IR::Vector<IR::Argument>* args,
+                const ExecutionState& state, SmallStepEvaluator::Result& result) {
+             const auto* advanceExpr = args->at(0)->expression;
+
+             if (!SymbolicEnv::isSymbolicValue(advanceExpr)) {
+                 stepToSubexpr(advanceExpr, result, state,
+                               [call](const Continuation::Parameter* v) {
+                                   auto* clonedCall = call->clone();
+                                   auto* arguments = clonedCall->arguments->clone();
+                                   auto* arg = arguments->at(0)->clone();
+                                   arg->expression = v->param;
+                                   (*arguments)[0] = arg;
+                                   clonedCall->arguments = arguments;
+                                   return Continuation::Return(clonedCall);
+                               });
+                 return;
+             }
+
+             // There are two possibilities. Either the advance amount is a constant or a runtime
+             // expression. In the first case, we just need to retrieve the value from the constant.
+             AdvanceInfo condInfo{};
+             if (const auto* advanceConst = advanceExpr->to<IR::Constant>()) {
+                 auto advanceValue = advanceConst->asInt();
+                 // Calculate the conditions for a failed or successful advance, given the size.
+                 condInfo = calculateSuccessfulParserAdvance(state, advanceValue);
+             } else {
+                 // Check whether advance expression is tainted.
+                 // If that is the case, we have no control or idea how much the cursor can be
+                 // advanced.
+                 auto advanceIsTainted = state.hasTaint(advanceExpr);
+                 if (advanceIsTainted) {
+                     TESTGEN_UNIMPLEMENTED(
+                         "The advance expression of %1% is tainted. We can not predict how much "
+                         "this call will advance the parser cursor. Abort.",
+                         call);
+                 }
+                 // In the second case, where the advance amount is a runtime expression, we need to
+                 // invoke the solver.
+                 // The size of the advance expression should be smaller than the maximum packet
+                 // size.
+                 auto* sizeRestriction = new IR::Leq(
+                     advanceExpr, IRUtils::getConstant(advanceExpr->type,
+                                                       ExecutionState::getMaxPacketLength_bits()));
+                 // The advance expression should ideally fit into a multiple of 8 bits.
+                 auto* bytesRestriction = new IR::Equ(
```

Comment:
I don't see why that's true.

Reply:
https://github.com/p4lang/behavioral-model/issues/727#issuecomment-470648582

It is the case for BMv2 only, but we have this as a general restriction until we specialize the advance for targets.